### PR TITLE
[autoscaler] Add documentation for multi node type autoscaling

### PR DESCRIPTION
**Pull request by ericl**

### CODE REVIEW

File: python/ray/autoscaler/resource_demand_scheduler.py [hunk 1]
```diff
@@ -30,18 +39,43 @@ def __init__(self, provider: NodeProvider,
         self.node_types = node_types
         self.max_workers = max_workers
 
-    def debug_string(self, nodes: List[NodeID],
-                     pending_nodes: Dict[NodeID, int]) -> str:
+    # TODO(ekl) take into account existing utilization of node resources. We
```
Comment (ericl) [inline python/ray/autoscaler/resource_demand_scheduler.py:42]: cc @wuisawesome

File: doc/source/cluster/autoscaling.rst [hunk 2]
```diff
@@ -0,0 +1,111 @@
+.. _ref-autoscaling:
+
+Cluster Autoscaling
+===================
+
+Basics
+------
+
+The Ray Cluster Launcher will automatically enable a load-based autoscaler. When cluster resource usage exceeds a configurable threshold (80% by default), new nodes will be launched up the specified ``max_workers`` limit (in the cluster config). When nodes are idle for more than a timeout, they will be removed, down to the ``min_workers`` limit. The head node is never removed.
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:9]:
```suggestion
The Ray Cluster Launcher will automatically enable a load-based autoscaler. When cluster resource usage exceeds a configurable threshold (80% by default), new nodes will be launched up to the specified ``max_workers`` limit (specified in the cluster config). When nodes are idle for more than a timeout, they will be removed, down to the ``min_workers`` limit. The head node is never removed.
```

File: doc/source/cluster/autoscaling.rst [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    # The minimum number of workers nodes to launch in addition to the head
+    # node. This number should be >= 0.
+    min_workers: 0
+
+    # The autoscaler will scale up the cluster to this target fraction of resource
+    # usage. For example, if a cluster of 10 nodes is 100% busy and
+    # target_utilization is 0.8, it would resize the cluster to 13. This fraction
+    # can be decreased to increase the aggressiveness of upscaling.
+    # This max value allowed is 1.0, which is the most conservative setting.
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:28]:
```suggestion
    # The max value allowed is 1.0, which is the most conservative setting.
```

File: doc/source/cluster/autoscaling.rst [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+.. code::
+
+    # An unique identifier for the head node and workers of this cluster.
+    cluster_name: default
+
+    # The minimum number of workers nodes to launch in addition to the head
+    # node. This number should be >= 0.
+    min_workers: 0
+
+    # The autoscaler will scale up the cluster to this target fraction of resource
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:24]:
```suggestion
    # The autoscaler will scale up the cluster to this target fraction of CPU
```
Not sure if this is accurate.

File: doc/source/cluster/autoscaling.rst [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    # usage. For example, if a cluster of 10 nodes is 100% busy and
+    # target_utilization is 0.8, it would resize the cluster to 13. This fraction
+    # can be decreased to increase the aggressiveness of upscaling.
+    # This max value allowed is 1.0, which is the most conservative setting.
+    target_utilization_fraction: 0.8
+
+    # If a node is idle for this many minutes, it will be removed.
+    idle_timeout_minutes: 5
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:32]: Add a note about what happens if there are actors or objects on that node?

File: doc/source/cluster/autoscaling.rst [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+Multiple Node Type Autoscaling
+------------------------------
+
+In 1.0, Ray supports multiple cluster node types. In this mode of operation, the scheduler will look at the queue of resource shape demands from the cluster (e.g., there might be 10 tasks queued each requesting ``{"GPU": 4, "CPU": 16}``), and specifically tries to add nodes that can fulfill these resource demands. This enables precise, rapid scale up as the autoscaler has more visibility into the backlog of work and resource shapes.
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:37]:
```suggestion
Ray supports multiple node types in a single cluster. In this mode of operation, the scheduler will look at the queue of resource shape demands from the cluster (e.g., there might be 10 tasks queued each requesting ``{"GPU": 4, "CPU": 16}``), and tries to add the minimum set of nodes that can fulfill these resource demands. This enables precise, rapid scale up compared to looking only at resource utilization, as the autoscaler also has visiblity into the aggregate resource load.
```

File: doc/source/cluster/autoscaling.rst [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+The concept of a cluster node type encompasses both the physical instance type (e.g., AWS p3.8xl GPU nodes vs m4.16xl CPU nodes), as well as other attributes (e.g., IAM role, the machine image, etc). Custom resources can be specified for each node type so that Ray is aware of the demand for specific node types at the application level (e.g., a task may request to be placed on a machine with a specific role or machine image via custom resource).
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:39]: Add a link to custom resources documentation?

File: doc/source/cluster/autoscaling.rst [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+Multi node type autoscaling operates in conjunction with the basic autoscaler. You may want to configure the basic autoscaler accordingly to act convervatively (i.e., set ``target_utilization_fraction: 1.0``).
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:41]:
```suggestion
Multi-node type autoscaling operates in conjunction with the basic autoscaler. You may want to configure the basic autoscaler accordingly to act conservatively (i.e., set ``target_utilization_fraction: 1.0``).
```

Can you add some reasoning here?

File: doc/source/cluster/autoscaling.rst [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+An example of configuring multiple node types is as follows `(full example) <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-multi-node-type.yaml>`__:
+
+.. code::
+
+    # Tell the autoscaler the allowed node types and the resources they provide.
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:47]:
```suggestion
    # Specify the allowed node types and the resources they provide.
```

File: doc/source/cluster/autoscaling.rst (hunk 4 above)
Comment (ericl) [reply doc/source/cluster/autoscaling.rst:24]: It's actually resource usage, not CPU

File: python/ray/autoscaler/aws/example-full.yaml [hunk 10]
```diff
@@ -42,7 +42,7 @@ docker:
 # usage. For example, if a cluster of 10 nodes is 100% busy and
 # target_utilization is 0.8, it would resize the cluster to 13. This fraction
 # can be decreased to increase the aggressiveness of upscaling.
-# This value must be less than 1.0 for scaling to happen.
+# This max value allowed is 1.0, which is the most conservative setting.
```
Comment (stephanie-wang) [inline python/ray/autoscaler/aws/example-full.yaml:45]:
```suggestion
# The max value allowed is 1.0, which is the most conservative setting.
```

File: doc/source/cluster/autoscaling.rst (hunk 4 above)
Comment (stephanie-wang) [reply doc/source/cluster/autoscaling.rst:24]: Could we specify what resources are considered?

File: doc/source/cluster/autoscaling.rst [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    # The autoscaler will scale up the cluster to this target fraction of resource
+    # usage. For example, if a cluster of 10 nodes is 100% busy and
+    # target_utilization is 0.8, it would resize the cluster to 13. This fraction
+    # can be decreased to increase the aggressiveness of upscaling.
+    # The max value allowed is 1.0, which is the most conservative setting.
+    target_utilization_fraction: 0.8
+
+    # If a node is idle for this many minutes, it will be removed. A node is
+    # considered idle if there are any tasks or actors running on it.
```
Comment (stephanie-wang) [inline doc/source/cluster/autoscaling.rst:32]:
```suggestion
    # considered idle if there are no tasks or actors running on it.
```

**Review by stephanie-wang — APPROVED** [review]

File: doc/source/cluster/autoscaling.rst (hunk 4 above)
Comment (ericl) [reply doc/source/cluster/autoscaling.rst:24]: I've updated this in the paragraph above.

**Pull request merged**